proposed/2022/DotnetUpdatePackage.md

```diff
@@ -0,0 +1,179 @@
+
+![](../../meta/resources/dotnetupdatepackage/dotnetupdatepackage.png)
+
+```
+dotnet update [<PROJECT>|<SOLUTION>] package <PACKAGE_NAME>
+    [-f|--framework <FRAMEWORK>] [--interactive]
+    [-n|--no-restore] [--package-directory <PACKAGE_DIRECTORY>]
+    [--prerelease] [-s|--source <SOURCE>] [-v|--version <VERSION>]
+    [--dry-run] [--highest-minor] [--highest-patch]
+
+dotnet update package -h|--help
+```
+
+#### Arguments
+
+- PROJECT | SOLUTION
+
+The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution or project is found, an error is thrown.
+
+- PACKAGE_NAME
+
+The package reference to add. Package name wildcards should be supported and update a subset of a glob-style pattern package name. i.e. `Microsoft.*` would update Microsoft packages. The commonly used symbols to support are:
+
+- `*` - zero or more characters.
+- `?` - single occurrence of any character.
+- `.` - literal "." character
+
+#### Options
+
+- -f|--framework <FRAMEWORK>
+
+Adds a package reference only when targeting a specific framework.
+
+- -?|-h|--help
+
+Prints out a description of how to use the command.
+
+- --interactive
+
+Allows the command to stop and wait for user input or action. For example, to complete authentication.
+
+- -n|--no-restore
+
+Adds a package reference without performing a restore preview and compatibility check.
+
+- --package-directory <PACKAGE_DIRECTORY>
+
+The directory where to restore the packages. The default package restore location is %userprofile%\.nuget\packages on Windows and ~/.nuget/packages on macOS and Linux. For more information, see Managing the global packages, cache, and temp folders in NuGet.
+
+- --prerelease
+
+Allows prerelease packages to be installed. Available since .NET Core 5 SDK
+
+- -s|--source <SOURCE>
+
+The URI of the NuGet package source to use during the restore operation.
+
+- -v|--version <VERSION>
+
+Version of the package. See NuGet package versioning.
+
+- --dry-run
+
+Displays what would be updated, but doesn't actually do the operation.
+
+- --highest-minor
+
+Considers only the packages with a matching major version number when searching for updated packages.
+
+- --highest-patch
+
+Considers only the packages with a matching major and minor version numbers when searching for updated packages.
+
+#### Examples
+
+- Updates all NuGet packages in the current working directory.
+
+```
+dotnet update package
+```
+
+- Updates a specific NuGet package in the current working directory.
+
+```
+dotnet update package <PACKAGE_NAME>
+```
+
+- Updates a glob of NuGet package(s) in the current working directory.
+
+```
+dotnet update package Microsoft.*
+```
+
+- Displays what would be updated in the update operation for the current working directory.
+
+```
+dotnet update package --dry-run
+```
+
+#### Exit Codes
```

Comment:
Is it all or nothing? Is it possible from some packages to update while other fail?

Reply:
Still thinking through it. Any suggestions?

Reply:
IMO, it shouldn't be all or nothing by default - but may want to provide an option for it if users want to take advantage of the exit code - but that's a secondary need.

If I have a project with an older framework where the latest version of half my packages still support my TFM, but the latest version of the other half don't, an all or nothing design would prevent me from bulk updating altogether.

Based on feedback about version preferences from the HaTS survey, I think the typical user will likely want to update all compatible packages to the latest stable version, even if other will fail. I think it's likely a fairly small subset that will take advantage of the exit code for automation purposes.

Reply:
I think `update by tfm compatibility` is a special scenario. 
It's one of those few scenarios where a failure is that exact package only. 

The trickier one is what if the latest versions causes resolution conflicts, such as downgrades etc. 
It becomes a user decision what they'd want to do then, so something semi interactive.  
`--no-restore` is a an option that *skips* validations.
Right now the most common failures at, NU1605, NU1105, NU1201 and NU1104.

Back to the compat scenario, I think it's the one that we're most likely to solve. Worth noting that we'd have performance/implementation challenges for any source that isn't nuget.org.
Right now, search by tfm is a search thing only, we'd need to validate the perf issues. 

tldr; I feel the pain of the customers, I run into some of these myself, but solving this in the first iteration might be a reach.

Reply:
Another "fun" case when updating by tfm compatibility that I run in to is that dotnet-ef supports something like dotnet core 2.1 in the latest version so if you update everything to the latest compatible version the EF Core packages might get out-of-sync. I really whish to be to update all packages to the latest compatible version but it's not as easy as it might seem.